modules/pipeline/providers/reconciler/task_reconciler.go

```diff
@@ -0,0 +1,391 @@
+			}
+		}
+
+		// teardown
+		tr.TeardownAfterReconcileDone(ctx, p, task)
+
+		// all done, exit
+		return rutil.NoNeedContinueWorking
+
+	}, rutil.WithContinueWorkingDefaultRetryInterval(tr.pr.r.Cfg.RetryInterval))
+}
+
+func (tr *defaultTaskReconciler) NeedReconcile(ctx context.Context, p *spec.Pipeline, task *spec.PipelineTask) bool {
+	return !task.Status.IsEndStatus()
+}
+
+func (tr *defaultTaskReconciler) SaveTaskIdempotently(ctx context.Context, p *spec.Pipeline, task *spec.PipelineTask) error {
+	if task.ID > 0 {
+		return nil
+	}
+
+	lastSuccessTaskMap, err := tr.pr.r.Cache.GetOrSetPipelineRerunSuccessTasksFromContext(p.ID)
+	if err != nil {
+		return err
+	}
+
+	stages, err := tr.pr.r.Cache.GetOrSetStagesFromContext(p.ID)
+	if err != nil {
+		return err
+	}
+
+	lastSuccessTask, ok := lastSuccessTaskMap[task.Name]
+	if ok {
+		task = lastSuccessTask
+		task.ID = 0
+		task.PipelineID = p.ID
+		task.StageID = stages[task.Extra.StageOrder].ID
+	}
+
+	err = tr.policy.AdaptPolicy(ctx, task)
+	if err != nil {
+		return err
+	}
+
+	// save task
+	if err := tr.pr.r.dbClient.CreatePipelineTask(task); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (tr *defaultTaskReconciler) ReconcileSnippetTask(ctx context.Context, p *spec.Pipeline, task *spec.PipelineTask) (err error) {
+	// generate corresponding snippet pipeline
+	var snippetPipeline *spec.Pipeline
+	if task.SnippetPipelineID != nil && *task.SnippetPipelineID > 0 {
+		snippetPipelineValue, err := tr.pr.r.dbClient.GetPipeline(task.SnippetPipelineID)
+		if err != nil {
+			return err
+		}
+		snippetPipeline = &snippetPipelineValue
+	} else {
+		snippetPipeline, err = tr.CreateSnippetPipeline(ctx, p, task)
+		if err != nil {
+			return err
+		}
+	}
+
+	// prepare before reconcile snippet pipeline
+	if err := tr.PrepareBeforeReconcileSnippetPipeline(ctx, snippetPipeline, task); err != nil {
+		return err
+	}
+
+	// reconcile as pipeline
+	// task status will be updated when snippet pipeline teardown
+	tr.pr.r.ReconcileOnePipeline(ctx, snippetPipeline.ID)
+
+	// setup snippet task info according to snippet pipeline result
+	if err := tr.fulfillParentSnippetTask(snippetPipeline); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (tr *defaultTaskReconciler) ReconcileNormalTask(ctx context.Context, p *spec.Pipeline, task *spec.PipelineTask) error {
+	var platformErrRetryTimes int
+	var onceCorrected bool
+	rutil.ContinueWorking(ctx, tr.log, func(ctx context.Context) rutil.WaitDuration {
+		// get executor
+		executor, err := actionexecutor.GetManager().Get(types.Name(task.GetExecutorName()))
+		if err != nil {
+			tr.log.Errorf("failed to get task executor(auto retry), err: %v", err)
+			return rutil.ContinueWorkingWithCustomInterval(tr.pr.r.Cfg.RetryInterval)
+		}
+
+		// generate framework to run task
+		framework := taskrun.New(ctx, task, executor, p, tr.pr.r.bdl, tr.pr.r.dbClient, tr.pr.r.actionAgentSvc, tr.pr.r.extMarketSvc, tr.pr.r.ClusterInfo)
```

Comment:
ugly params, can you refactor to a better New?

Reply:
taskrun pkg will be refactored totally later.